okapi-core/src/main/java/org/folio/okapi/managers/InternalModule.java

```diff
@@ -768,154 +729,121 @@ private void installModulesForTenant(ProxyContext pc, String id,
           TenantModuleDescriptor[].class);
       List<TenantModuleDescriptor> tm = new LinkedList<>();
       Collections.addAll(tm, tml);
-      tenantManager.installUpgradeModules(id, pc, options, tm, res -> {
-        if (res.failed()) {
-          fut.handle(new Failure<>(res.getType(), res.cause()));
-        } else {
-          logger.info("installUpgradeModules returns: {}", Json.encodePrettily(res.result()));
-          fut.handle(new Success<>(Json.encodePrettily(res.result())));
-        }
-      });
+      UUID installId = UUID.randomUUID();
+      return tenantManager.installUpgradeCreate(tenantId, installId.toString(), pc, options, tm)
+          .compose(res -> {
+            String jsonResponse = Json.encodePrettily(res);
+            logger.info("installTenantModulesPost returns: {}", jsonResponse);
+            if (options.getAsync()) {
+              return location(pc, installId.toString(), null, jsonResponse);
+            } else {
+              return Future.succeededFuture(jsonResponse);
+            }
+          });
     } catch (DecodeException ex) {
-      fut.handle(new Failure<>(ErrorType.USER, ex));
+      return Future.failedFuture(new OkapiError(ErrorType.USER, ex.getMessage()));
     }
   }
 
-  private void upgradeModulesForTenant(ProxyContext pc, String id,
-                                       Handler<ExtendedAsyncResult<String>> fut) {
+  private Future<String> installTenantModulesGetList(String tenantId) {
+    return tenantManager.installUpgradeGetList(tenantId)
+        .compose(installJobList -> Future.succeededFuture(Json.encodePrettily(installJobList)));
+  }
+
+  private Future<String> installTenantModulesGet(String tenantId, String installId) {
+    return tenantManager.installUpgradeGet(tenantId, installId)
+        .compose(installJob -> Future.succeededFuture(Json.encodePrettily(installJob)));
+  }
+
+  private Future<String> upgradeModulesForTenant(ProxyContext pc, String id) {
 
     TenantInstallOptions options = ModuleUtil.createTenantOptions(pc.getCtx().request());
-    tenantManager.installUpgradeModules(id, pc, options, null, res -> {
-      if (res.failed()) {
-        fut.handle(new Failure<>(res.getType(), res.cause()));
-      } else {
-        logger.info("installUpgradeModules returns: {}", Json.encodePrettily(res.result()));
-        fut.handle(new Success<>(Json.encodePrettily(res.result())));
-      }
-    });
+    UUID installId = UUID.randomUUID();
+    return tenantManager.installUpgradeCreate(id, installId.toString(), pc, options, null)
+        .compose(res -> Future.succeededFuture(Json.encodePrettily(res)));
```

Comment:
Do we support checking the status of async upgrade?

Reply:
yes. Should probably be mentioned.

Reply:
This `/upgrade` endpoint method does not return `location` header, so how does the caller know where to track the `status`?